File: jaas-jwt-samples/php/jaas-jwt.php
#!/usr/bin/php
<?php

require __DIR__ . '/vendor/autoload.php';

/**
 * PHP code below is used to generate a JaaS JWT.
 * You can copy the code below in your implementation.
 */
use Jose\Component\Core\AlgorithmManager;
use Jose\Component\Core\JWK;
use Jose\Component\Signature\Algorithm\RS256;
use Jose\Component\Signature\JWSBuilder;
use Jose\Component\KeyManagement\JWKFactory;
use Jose\Component\Signature\Serializer\CompactSerializer;

/**
 * Change the variables below.
 */
$API_KEY="my api key";
$APP_ID="my AppID"; // Your AppID (previously tenant)
$USER_EMAIL="myemail@email.com";
$USER_NAME="my user name";
$USER_IS_MODERATOR=true;
$USER_AVATAR_URL="";
$USER_ID="b2c94a50-e53b-4afc-8bef-3132f3ec27dc";
$LIVESTREAMING_IS_ENABLED=false;
$RECORDING_IS_ENABLED=true;
$OUTBOUND_IS_ENABLED=false;
$TRANSCRIPTION_IS_ENABLED=false;
$EXP_DELAY_SEC=7200;
$NBF_DELAY_SEC=10;
///

/**
 * We read the JSON Web Key (https://tools.ietf.org/html/rfc7517) 
 * from the private key we generated at https://jaas.8x8.vc/#/apikeys .
 * 
 * @var \Jose\Component\Core\JWK jwk
 */
$jwk = JWKFactory::createFromKeyFile("./rsa-private.key");

/**
 * Setup the algoritm used to sign the token.
 * @var \Jose\Component\Core\AlgorithmManager $algorithm
 */
$algorithm = new AlgorithmManager([
    new RS256()
]);

/**
 * The builder will create and sign the token.
 * @var \Jose\Component\Signature\JWSBuilder $jwsBuilder
 */
$jwsBuilder = new JWSBuilder($algorithm);

/**
 * Must setup JaaS payload!
 * Change the claims below or using the variables from above!
 */
$payload = json_encode([
    'iss' => 'chat',
    'aud' => 'jitsi',
    'exp' => time() + $EXP_DELAY_SEC,
    'nbf' => time() - $NBF_DELAY_SEC,
    'room'=> '*',
    'sub' => $APP_ID,
    'context' => [
        'user' => [
            'moderator' => $USER_IS_MODERATOR ? "true" : "false",
            'email' => $USER_EMAIL,
            'name' => $USER_NAME,
            'avatar' => $USER_AVATAR_URL,
            'id' => $USER_ID
        ],
        'features' => [
            'recording' => $RECORDING_IS_ENABLED ? "true" : "false",
            'livestreaming' => $LIVESTREAMING_IS_ENABLED ? "true" : "false",
            'transcription' => $TRANSCRIPTION_IS_ENABLED ? "true" : "false",
            'outbound-call' => $OUTBOUND_IS_ENABLED ? "true" : "false"
        ]
    ]
]);

/**
 * Create a JSON Web Signature (https://tools.ietf.org/html/rfc7515)
 * using the payload created above and the api key specified for the kid claim.
 * 'alg' (RS256) and 'typ' claims are also needed.
 */
$jws = $jwsBuilder
        ->create()
        ->withPayload($payload)
        ->addSignature($jwk, [
            'alg' => 'RS256',
            'kid' => $API_KEY,
            'typ' => 'JWT'
        ])
        ->build();

/**
 * We use the serializer to base64 encode into the final token.
 * @var \Jose\Component\Signature\Serializer\CompactSerializer $serializer
 */
$serializer = new CompactSerializer();
$token = $serializer->serialize($jws, 0);

/**
 * Write the token to standard output.
 */
echo $token;

File: jaas-jwt-samples/ruby/jaas-jwt.rb
require "openssl"
require "jwt"

API_KEY="my api key" # See https://jaas.8x8.vc/#/apikeys for more info.
APP_ID="my AppID" # Sets the AppID / tenant
USER_NAME="my user name" # Sets the user name
USER_EMAIL="my user email" # Sets the user email
USER_AVATAR_URL="my avatar url" # Sets the users avatar url
USER_ID="my user id" # Set a unique id for the user, corresponding to your backend
USER_IS_MODERATOR=false # Sets the user as moderator
RECORDING_ENABLED=false # Sets recording feature enabled
LIVESTREAMING_ENABLED=false # Sets live steraming feature enabled
OUTBOUND_CALL_ENABLED=false # Sets outbound call feature enabled
TRANSCRIPTION_ENABLED=false # Sets transcription feature enabled
ROOM_NAME="my room name or *" # Sets the room name
TIME_NOW=Time.now.to_i # Sets to current local time
EXP_TIME_DELAY=7200 # No real need to modify this
NBF_TIME_DELAY=10 # No real need to modify this

PAYLOAD = {
  "aud": "jitsi",
  "exp": TIME_NOW + EXP_TIME_DELAY,
  "nbf": TIME_NOW - NBF_TIME_DELAY,
  "iss": "chat",
  "room": ROOM_NAME,
  "sub": APP_ID,

  "context": {
    "features": {
      "livestreaming": LIVESTREAMING_ENABLED,
      "outbound-call": OUTBOUND_CALL_ENABLED,
      "transcription": TRANSCRIPTION_ENABLED,
      "recording": RECORDING_ENABLED
    },
    "user": {
      "moderator": USER_IS_MODERATOR,
      "name": USER_NAME,
      "id": USER_ID,
      "avatar": USER_AVATAR_URL,
      "email": USER_EMAIL
    }
  }
}

private_key = OpenSSL::PKey::RSA.new File.read("/path/to/your_private_key.pk")
jwt = JWT.encode PAYLOAD, private_key, "RS256", {"typ": "JWT", "kid": API_KEY}
